mas_perception_libs/ros/src/mas_perception_libs/scene_detection_action.py

```diff
@@ -24,15 +33,133 @@ def _execute_cb(self, goal):
+            self._action_server.set_aborted(text=e.message)
+            return
+
+        rospy.loginfo('creating action result and setting success')
+        result = ImageDetectionActionServer._get_action_result(transformed_cloud_msg, bounding_boxes,
+                                                               classes, confidences)
         self._action_server.set_succeeded(result)
+
+    def _cloud_callback(self, cloud_msg):
+        self._cloud_msg = cloud_msg
+
+    def _transform_cloud(self, cloud_msg):
+        try:
+            common_time = self._tf_listener.getLatestCommonTime(self._target_frame, cloud_msg.header.frame_id)
+            cloud_msg.header.stamp = common_time
+            self._tf_listener.waitForTransform(self._target_frame, cloud_msg.header.frame_id,
+                                               cloud_msg.header.stamp, rospy.Duration(1))
+            tf_matrix = self._tf_listener.asMatrix(self._target_frame, cloud_msg.header)
+        except(tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
+            raise RuntimeError('Unable to transform {0} -> {1}'.format(cloud_msg.header.frame_id, self._target_frame))
+
+        return transform_point_cloud(cloud_msg, tf_matrix, self._target_frame)
+
+    @staticmethod
+    def _get_action_result(cloud_msg, bounding_boxes, classes, confidences):
+        """
+        :type cloud_msg: PointCloud2
+        :type bounding_boxes: list
+        :type classes: list
+        :type confidences: list
+        :rtype: DetectSceneResult
+        """
+        result = DetectSceneResult()
+        plane = Plane()
+        for index, box in enumerate(bounding_boxes):
+            detected_obj = Object()
+            detected_obj.name = classes[index]
+            detected_obj.probability = confidences[index]
+            cropped_cloud = crop_organized_cloud_msg(cloud_msg, box)
+            detected_obj.pointcloud = cropped_cloud
+            detected_obj.rgb_image = cloud_msg_to_image_msg(cropped_cloud)
+            detected_obj.pose = ImageDetectionActionServer._estimate_grasp_pose(cropped_cloud)
+            plane.object_list.objects.append(detected_obj)
+        result.planes.append(plane)
+        return result
+
+    @staticmethod
+    def _estimate_grasp_pose(cropped_cloud):
+        """
+        :type cropped_cloud: PointCloud2
+        :rtype: PoseStamped
+        """
+        # TODO(minhnh) make this configurable
+        pose_stamped = PoseStamped()
+        pose_stamped.header = cropped_cloud.header
+        # TODO(minhnh) make cloud_to_mat() instead of crop_cloud_to_xyz
+        cropped_coord = crop_cloud_to_xyz(cropped_cloud,
+                                          BoundingBox2D((0, 0, cropped_cloud.width, cropped_cloud.height)))
+        mean_coord = np.nanmean(np.reshape(cropped_coord, (-1, 3)), axis=0)
+        min_coord = np.nanmin(np.reshape(cropped_coord, (-1, 3)), axis=0)
+        pose_stamped.pose.position.x = min_coord[0]
```

Comment:
can you describe here how the pose will look for an object? i.e. which directions are y and z? 
Are these assumptions valid for different types of grasps (top grasp, side grasp etc.)

Reply:
I removed this part for now since it overlaps with pose calculation in `ObjectDetector`. I will think on this a bit more and make a different PR.